test/unit/org/apache/cassandra/inject/Injections.java

```diff
@@ -0,0 +1,623 @@
+         * Adds a new invoke point to the injection. Adding a new invoke point cause creation of a new rule because
+         * a single rule can only have a single invoke point.
+         */
+        public B add(InvokePointBuilder ipb)
+        {
+            invokePointBuilders.add(ipb);
+            return (B) this;
+        }
+    }
+
+    public abstract static class CrossProductInjectionBuilder<T extends Injection, B extends InjectionBuilder<T>>
+            extends MultiInvokePointInjectionBuilder<T, B>
+    {
+        final LinkedList<ActionBuilder> actionBuilders = new LinkedList<>();
+
+        CrossProductInjectionBuilder(String id)
+        {
+            super(id);
+        }
+
+        Rule[] getRules()
+        {
+            Rule[] rules = new Rule[actionBuilders.size() * invokePointBuilders.size()];
+            int i = 0;
+            for (InvokePointBuilder ipb : invokePointBuilders)
+            {
+                for (ActionBuilder ab : actionBuilders)
+                {
+                    rules[i++] = Rule.newRule(id + "_" + i, ab, ipb);
+                }
+            }
+            return rules;
+        }
+
+        /**
+         * Adds a new action to the injection. Adding a new action cause creation of a new rule because a single rule
+         * can only have a single action. Do not confuse an action with a statement - an action is bundle of bindings,
+         * condition under which it can be invoked and a sequence of statements.
+         *
+         * If you just need to add a statement to the existing action, see {@link #withLastActionBuilder(Consumer)} and
+         * {@link #lastActionBuilder()}.
+         */
+        public B add(ActionBuilder builder)
+        {
+            if (builder.getHelperClass() == null)
+            {
+                builder.withHelperClass(Helper.class);
+            }
+            builder.conditions().when(getIsEnabledExpression());
+            actionBuilders.add(builder);
+            return (B) this;
+        }
+
+        /**
+         * @see #add(ActionBuilder)
+         */
+        public B add(ActionBuilder.Builder builder)
+        {
+            return add(builder.toActionBuilder());
+        }
+
+        /**
+         * Allows to modify the last defined action. You can add new bindings, conditions and statements. If you need
+         * to create a new action, please see {@link #add(ActionBuilder)}.
+         */
+        public ActionBuilder lastActionBuilder()
+        {
+            Preconditions.checkState(!actionBuilders.isEmpty());
+            ActionBuilder ab = actionBuilders.getLast();
+            return ab;
+        }
+
+        /**
+         * @see #lastActionBuilder()
+         */
+        public B withLastActionBuilder(Consumer<ActionBuilder> builder)
+        {
+            Preconditions.checkState(!actionBuilders.isEmpty());
+            ActionBuilder ab = actionBuilders.getLast();
+            builder.accept(ab);
+            return (B) this;
```

Comment:
Nit: Add `@SuppressWarnings("unchecked")`

Reply:
Same as above. The method is never used.